src/storage/garbage_collector.cpp

```diff
@@ -87,32 +122,111 @@ uint32_t GarbageCollector::ProcessUnlinkQueue() {
+        buffers_to_deallocate_.push_front(buf);
+      } else {
+        // We didn't unlink all of the UndoRecords (UnlinkUndoRecord returned false due to a write-write conflict),
+        // requeue buf for next GC run. Unlinked UndoRecords will be skipped on the next time around since we use the
+        // table pointer of an UndoRecord as the internal marker of being unlinked or not
+        buf_requeue.push_front(buf);
+      }
     }
   }
 
-  // Requeue any txns that we were still visible to running transactions
-  txns_to_unlink_ = transaction::TransactionQueue(std::move(requeue));
+  // Requeue any txns that have an undo record still visible to some running transaction
+  if (!requeue.empty()) {
+    txns_to_unlink_ = transaction::TransactionQueue(std::move(requeue));
+  }
+
+  // Requeue any compaction buffers that that have an undo record still visible to some running transaction
+  if (!buf_requeue.empty()) {
+    buffers_to_unlink_ = std::forward_list<storage::UndoBuffer *>(std::move(buf_requeue));
+  }
 
   return txns_processed;
 }
 
+bool GarbageCollector::ProcessUndoRecord(UndoRecord *const undo_record,
+                                         std::vector<transaction::timestamp_t> *const active_txns,
+                                         std::unordered_set<TupleSlot> *visited_slots) {
+  DataTable *table = undo_record->Table();
+  // If this UndoRecord has already been processed, we can skip it
+  if (undo_record->txnptr_.IsNull()) return true;
+  const TupleSlot slot = undo_record->Slot();
+  // Process this tuple only
+  if (visited_slots->insert(slot).second) {
+    // Perform interval gc for the entire version chain excluding the head of the chain
+    ProcessTupleVersionChain(table, slot, active_txns);
+    ProcessTupleVersionChainHead(table, slot, active_txns);
+  }
+
+  return undo_record->txnptr_.IsNull();
+}
+
+void GarbageCollector::ProcessTupleVersionChainHead(DataTable *const table, TupleSlot slot,
+                                                    std::vector<transaction::timestamp_t> *const active_txns) {
+  UndoRecord *version_chain_head;
+  const TupleAccessStrategy &accessor = table->accessor_;
+  version_chain_head = table->AtomicallyReadVersionPtr(slot, accessor);
+  if (version_chain_head == nullptr) {
+    // This version chain is empty, so we can skip it
+    return;
+  }
+  // Perform gc for head of the chain
+  // Assuming can garbage collect any version greater than the oldest timestamp
+  transaction::timestamp_t version_ptr_timestamp = version_chain_head->Timestamp().load();
+  // If there are no active transactions, or if the version pointer is older than the oldest active transaction,
+  // Collect the head of the chain using compare and swap
+  // Note that active_txns is sorted in descending order, so its tail should have the oldest txn's timestamp
+  if (version_ptr_timestamp < active_txns->back()) {
```

Comment:
Calling `back` on empty vector can cause undefined behaviour, need to check with `ative_txns->empty()` first

Reply:
Transaction Manager ensures that if the active_txns is not empty, it inserts a timestamp.

Reply:
`GetActiveTxns` is based on the function `OldestTransactionStartTime`, which was used by the original GC. `OldestTransactionStartTime` returns the oldest active transaction's start time. If there are no active transactions, it returns the current time (that is, it returns `time_`, the internal counter). Similar to `OldestTransactionStartTime`, `GetActiveTxns` returns the active transactions' timestamps. And if there are none, it returns the current time.